src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs

```diff
@@ -94,6 +95,7 @@ public static class ExpressionMetadataProvider
                 //    m => foo (arbitrary expression)
                 //    m => m.Widgets[0] (expression ending with non-property-access)
                 metadata = metadataProvider.GetMetadataForType(typeof(TResult));
+                Debug.Assert(metadata != null);
```

Comment:
It makes it clear what the expectations are on both branches wrt the null-ness of `metadata`. Zoom out and look at the whole block.